### PR TITLE
Pointwise fuse cat with pointwise inputs or outputs and <= 4 inputs

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #111233

Improves perf of llama_v2 locally from 1.55 -> 1.57

The initial heuristic is to lower to pointwise if # of inputs is <= 4, and all the inputs are pointwise or cannot be memory planned away, or if all the outputs are pointwise. 

Perf run was +3% on inference.. There are definitely instances where we should be lowering to foreach_kernels, but it's less flexible for fusion. The motivating example was:

```
def rotate_half(x):
    """Rotates half the hidden dims of the input."""
    x1 = x[..., : x.shape[-1] // 2]
    x2 = x[..., x.shape[-1] // 2 :]
    return torch.cat((-x2, x1), dim=-1)

def apply_rotary_pos_emb(q, k, cos, sin):
    iota =  torch.ops.prims.iota.default(512, start = 0, step = 1, dtype = torch.int64, device = device(type='cuda', index=0), requires_grad = False)
    
    # File: /scratch/eellison/work/torchdynamo/lib/python3.8/site-packages/transformers/models/llama/modeling_llama.py:657, code: position_ids = position_ids.unsqueeze(0).view(-1, seq_length)
    unsqueeze = torch.ops.aten.unsqueeze.default(iota, 0)
    position_ids = torch.ops.aten.reshape.default(unsqueeze, [-1, 512]);  unsqueeze = None

    # The first two dimensions of cos and sin are always 1, so we can `squeeze` them.
    cos = cos.squeeze(1).squeeze(0)  # [seq_len, dim]
    sin = sin.squeeze(1).squeeze(0)  # [seq_len, dim]
    cos = cos[position_ids].unsqueeze(1)  # [bs, 1, seq_len, dim]
    sin = sin[position_ids].unsqueeze(1)  # [bs, 1, seq_len, dim]
    q_embed = (q * cos) + (rotate_half(q) * sin)
    k_embed = (k * cos) + (rotate_half(k) * sin)
    return q_embed, k_embed
```

Also not sure if I should be more worried about concatting reduction->pointwise inputs.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

**Review by Chillee — COMMENTED** [review]
Seems like there's some test failures, but the overall optimization makes a lot of sense to me :)

Also need some tests. I think this is a good usage of the "read/write" based tests in `test_perf.py`.

File: torch/_inductor/lowering.py [hunk 1]
```diff
@@ -966,6 +1036,31 @@ def cat(inputs, dim=0):
...
+    def should_lower_cat_input(x) -> bool:
+        if isinstance(x, TensorBox):
+            if isinstance(x.data, ir.BaseView):
+                return should_lower_cat_input(x.data.unwrap_view())
+            else:
+                return should_lower_cat_input(x.data)
+        if isinstance(x, ir.StorageBox):
+            return should_lower_cat_input(x.data)
+
+        # TODO: for input buffers, if we cant fuse use, consider foreach lowering
```
[the remark this answers is not in the document]
Comment (eellison) [reply torch/_inductor/lowering.py:1049]: Yea, I think the right heuristic is:

1. inputs are pointwise, fuse it
2. If the inputs are FixedLayout and cant be planned out:
- pointwise uses  fuse it
- non pointwise uses, foreach_kernel

**Review by jansel — CHANGES_REQUESTED** [review]
High level looks fine, but tests are failing.

File: test/inductor/test_perf.py [hunk 2]
```diff
@@ -343,6 +343,20 @@ def f(a, b, c):
...
         self.assertExpectedInline(count_numel(f, *inp), """500""")
 
+    def test_cat_pointwise_inps(self):
+        def f(a, b, c):
+            return torch.cat((a + 1, b + 2, c + 3)) + 10
+
+        inp = (T(10, 10), T(10, 10), T(10, 10))
+        self.assertExpectedInline(count_numel(f, *inp), """600""")
+
+    def test_cat_pointwise_inps_non_fused(self):
```
Comment (Chillee) [inline test/inductor/test_perf.py:353]: I think this kind of test is a little bit worse, since you could imagine in the future we decide to tune the heuristic and we end up fusing it.

I think it would be good to add some more tests for the case where this heuristic *could* trigger but it's suboptimal (i.e. inputs are from a non-fusible op)

File: torch/_inductor/lowering.py [hunk 3]
```diff
@@ -969,6 +1039,37 @@ def cat(inputs, dim=0):
...
+        if isinstance(x, ir.StorageBox):
+            return should_lower_cat_input(x.data)
+
+        if isinstance(x, ir.Pointwise):
+            return True
+
+        return False
+
+    # TODO - refine heuristics
+    POINTWISE_CAT_MAX_INPUTS = 4
```
Comment (Chillee) [inline torch/_inductor/lowering.py:1065]: We should probably make this a config.

File: torch/_inductor/lowering.py [hunk 4]
```diff
@@ -4198,7 +4299,6 @@ def loader(index, reduction_index):
         assert len(reduction_index) == len(reduced_idx)
         if keepdims:
             assert len(index) == len(size)
-            assert all(index[i] == 0 for i in reduced_idx)
```
Comment (Chillee) [inline torch/_inductor/lowering.py:4201]: I think an user recently reported this as a bug... is this something we're fixing?

File: test/inductor/test_perf.py [hunk 5]
```diff
@@ -343,6 +343,20 @@ def f(a, b, c):
         inp = (T(10, 10), T(10, 10), T(10, 10))
         self.assertExpectedInline(count_numel(f, *inp), """500""")
 
+    def test_cat_pointwise_inps(self):
```
Comment (Chillee) [inline test/inductor/test_perf.py:346]: If these are the only tests I want to see some more :P Cover the different cases.

**Review by Chillee — COMMENTED** [review]
You should also do some microbenchmarking of cases like
```
torch.cat([x: [N], torch.zeros(1)])
```

I wonder whether we'll need another heuristic to make sure the shapes are not too far apart from each other.

File: torch/_inductor/lowering.py (hunk 4 above)
Comment (eellison) [reply torch/_inductor/lowering.py:4201]: Yea, I confirmed with @peterbell10, this check doesn't make sense with masked loads

File: test/inductor/test_perf.py (hunk 5 above)
Comment (eellison) [reply test/inductor/test_perf.py:346]: There are many existing cat tests in this file. But I can add more.

File: test/inductor/test_perf.py (hunk 2 above)
Comment (eellison) [reply test/inductor/test_perf.py:353]: inputs are from a non-fusible op - you mean where we should be able to memory plan away the cat but we don't ?

Comment (Chillee) [reply test/inductor/test_perf.py]: Actually, I'm thinking about if we have something like `cat([mm(a), mm(b), cos(c)]) + 3`. In this case, I think there would be a total of 8 accesses without this optimization applying, and 6 if it did?

So perhaps it's worth it as long as the downstream operators are fusible?

**Review by Chillee — APPROVED** [review]

**Comment by Chillee** [comment]
Failure looks plausibly related though?

File: test/inductor/test_perf.py (hunk 2 above)
Comment (eellison) [reply test/inductor/test_perf.py]: For this case, we would actually apply it. That's the pointwise_uses check.

Comment (Chillee) [reply test/inductor/test_perf.py]: I think you want at least one pointwise input right?
